server/s3_copy_object_action.cc

```diff
@@ -234,74 +223,214 @@ void S3CopyObjectAction::validate_copyobject_request() {
 }
 
-// Create destination object
-void S3CopyObjectAction::create_object() {
+void S3CopyObjectAction::write_data_block() {
+  s3_log(S3_LOG_INFO, request_id, "Entering");
+
+  assert(!write_in_progress);
+  assert(bytes_left_to_read > 0);
+
+  S3BufferSequence buffer_sequence;
+
+  char* p_data;
+  size_t block_size = motr_reader->get_first_block(&p_data);
+
+  unsigned block_in_chunk = 0;
+  size_t bytes_in_chunk = 0;
+
+  while (block_size) {
+    assert(p_data != NULL);
+    assert(block_size && block_size <= (size_t)motr_unit_size);
+
+    if (block_size >= bytes_left_to_read) {
+      block_size = bytes_left_to_read;
+      bytes_left_to_read = 0;
+    } else {
+      bytes_left_to_read -= block_size;
+    }
+    buffer_sequence.emplace_back(p_data, block_size);
+
+    bytes_in_chunk += block_size;
+    ++block_in_chunk;
+
+    block_size = motr_reader->get_next_block(&p_data);
+  }
+  assert(block_in_chunk);
+  assert(bytes_in_chunk);
+
+  s3_log(S3_LOG_DEBUG, request_id, "Got %zu bytes in %u blocks", bytes_in_chunk,
+         block_in_chunk);
+
+  motr_writer->write_content(
+      std::bind(&S3CopyObjectAction::write_data_block_success, this),
+      std::bind(&S3CopyObjectAction::write_data_block_failed, this),
+      std::move(buffer_sequence), motr_unit_size);
+
+  write_in_progress = true;
+
+  s3_log(S3_LOG_DEBUG, NULL, "Exiting\n");
+}
+
+void S3CopyObjectAction::write_data_block_success() {
+  s3_log(S3_LOG_INFO, request_id, "Entering");
+
+  assert(write_in_progress);
+  write_in_progress = false;
+
+  if (bytes_left_to_read) {
+    read_data_block();
+  } else {
+    s3_put_action_state = S3PutObjectActionState::writeComplete;
+    next();
+  }
+  s3_log(S3_LOG_DEBUG, NULL, "Exiting\n");
+}
+
+void S3CopyObjectAction::write_data_block_failed() {
+  assert(write_in_progress);
+  s3_put_action_state = S3PutObjectActionState::writeFailed;
+
+  s3_log(S3_LOG_INFO, request_id, "Failed to write object data to motr");
+  set_s3_error("InternalError");
+
+  if (read_in_progress) {
+    copy_failed = true;
+  } else {
+    send_response_to_s3_client();
+  }
+  s3_log(S3_LOG_DEBUG, NULL, "Exiting\n");
+}
+
+void S3CopyObjectAction::save_metadata() {
   s3_log(S3_LOG_INFO, request_id, "Entering\n");
-  next();
+
+  // for shutdown testcases, check FI and set shutdown signal
+  S3_CHECK_FI_AND_SET_SHUTDOWN_SIGNAL("put_object_action_save_metadata_pass");
+
+  new_object_metadata->reset_date_time_to_current();
+  new_object_metadata->set_content_length(std::to_string(total_data_to_stream));
+  new_object_metadata->set_content_type(
+      source_object_metadata->get_content_type());
+  new_object_metadata->set_md5(motr_writer->get_content_md5());
+  // new_object_metadata->set_tags(new_object_tags_map);
+
+  for (auto it : request->get_in_headers_copy()) {
+    if (it.first.find("x-amz-meta-") != std::string::npos) {
+      s3_log(S3_LOG_DEBUG, request_id,
+             "Writing user metadata on object: [%s] -> [%s]\n",
```

Comment:
"writing source object metadata on destination object.." ?

Reply:
Metadata attributes are being overwritten here according HTTP headers of the request.

Reply:
Ok ok.. I got your point... but we need to read a header: **x-amz-metadata-directive** 

This header can have two values:
COPY (default)
REPLACE.
As of now in milestone 1, we are only supporting COPY value.

Reply:
With COPY, all source metadata is copied to destination object.

with REPLACE value, user can give metadata with copy-object request, and the server will replace metadata on destination with passed metadata values

Reply:
Clear. Thanks! Will update.